src/aics-image-viewer/shared/utils/storage.ts

```diff
@@ -0,0 +1,162 @@
+import { mapValues } from "lodash";
+
+import type { MetadataRecord } from "../types";
+
+const QUEUE_KEY = "GLOBAL_ENTRY_QUEUE";
+
+const enum StorageEntryType {
+  Scenes = "scenes",
+  Meta = "meta",
+}
+
+const MAX_ENTRIES: { [K in StorageEntryType]: number } = {
+  scenes: 25,
+  meta: 1000,
+};
+
+/**
+ * Wrapper for `window.localStorage.setItem` that accepts a `queue` of evictable storage keys, and attempts to safely
+ * handle exceeding the storage quota by removing items off the back of the `queue` until space is available.
+ *
+ * Also, unlike `setItem`, this function can accept `value`s of type `string[]`. This allows `queue` to also be passed
+ * into `value`, so that it can be written to storage *after* picking up any evictions that are made during the write.
+ */
+function setStorageItem(key: string, value: string | string[], queue: string[]): void {
+  let success = false;
+  while (!success) {
+    try {
+      window.localStorage.setItem(key, Array.isArray(value) ? value.join(",") : value);
+      success = true;
+    } catch (e) {
+      if (e instanceof DOMException && e.name === "QuotaExceededError") {
+        const evictKey = queue.shift();
+        if (evictKey === undefined) {
+          const err = new Error("Tried to insert a single entry into local storage that was larger than the quota.");
```

Comment:
How will users be notified about this? Is there an onscreen error message?

Reply:
For users it needs to be worded differently.  The wording is good for an engineer trying to debug